### PR TITLE
Issue #257 - support for optional test method parameters in TestCaseAttr...

### PR DESCRIPTION
**Pull request by hiredgunhouse**
...ibute.
Fix for issue #257. I finally got around to it, sorry for the delay. Any comments/suggestions are welcomed.

### CODE REVIEW

**Comment by rprouse** [comment]
This looks good to me. @oznetmaster, can you take a quick look to see if there is anything that raises CF issues? I am also going to do a pull and a quick test with these changes as a double check. The tests look good, but I want to write a test with default params in my test solution. Once I do that, I will merge.

**Comment by oznetmaster** [comment]
Not at computer right now.  I will be able to check this out this afternoon (my time), so I should have a response in about 5-6 hours.

**Comment by oznetmaster** [comment]
I am sorry, but I ended up being away all day.  I will look at this first thing in the morning, if you want to wait until then

**Comment by rprouse** [comment]
@oznetmaster, no worries, I will merge tomorrow after you get a chance to look at it. I haven't had a chance to run my tests yet either.

**Comment by CharliePoole** [comment]
You have only tested this under .NET 4.0 and 4.5. However, it's a C# feature and works when .NET 2.0 or 3.5 are targeted, provided C# 4.0 is being used, which it is in our case.

So, the new testdata file should be added to the 2.0 and 3.5 projects and the #if should be removed. The new tests should not be under the control of #if NET_4_0 || NET_4_5 either.

**Comment by CharliePoole** [comment]
Our CV builds are done using VS2008, so default parameters are not permitted. Both the testdata and the test entries should be controlled by #if !NETCF.

My assumption, since this is a compiler feature, is that it will work under the silverlight and portable builds. It's easy enough to check before doing the work by just adding a default value to some method in those builds.

@hiredgunhouse Our practice is to add each file to each build and to control inclusion by conditionals. It makes it easier to search for things. However, If you don't have a copy of VS2008 handy to open the CF solution, don't worry about it. One of us will add it the next time we are in there.

File: src/NUnitFramework/framework/Internal/Builders/NUnitTestCaseBuilder.cs [hunk 1]
```diff
@@ -136,7 +136,14 @@ private static bool CheckTestMethodSignature(TestMethod testMethod, ParameterSet
...
             parameters = testMethod.Method.GetParameters();
 
-            int argsNeeded = parameters.Length;
+            int minArgsNeeded = 0;
+            foreach (var parameter in parameters)
+            {
+                // IsOptional is supported since .NET 1.1 
+                if (!parameter.IsOptional)
+                    minArgsNeeded++;
+            }
```
Comment (CharliePoole) [inline src/NUnitFramework/framework/Internal/Builders/NUnitTestCaseBuilder.cs]: ParameterInfo.IsOptional is not available under CF 3.5. This whole sequence should be conditional, with minArgsNeeded set to the same as maxArgsNeeded under NETCF.

**Comment by oznetmaster** [comment]
I agree.

**Comment by hiredgunhouse** [comment]
I applied your comments, however I don't have VS2008 installed so I did the changes blindly (hope it builds). Also R#'s unit test runner says that the test for SL-50 and Portable projects are inconclusive cuase could not be run (I get an exception that System.Windows assembly is missing - version 5.0.5.x) - not sure yet how to fix that.

**Comment by CharliePoole** [comment]
No sweat. We'll test it in CF, Silverlight and Portable.


All those tests are self-executing. For example, for Silverlight, you just
set it as the startup project in VS and run it.

**Comment by hiredgunhouse** [comment]
I run the Silverlight and Portable and looks like my tests are passing.

**Comment by rprouse** [comment]
Looks good, If there are any additional CF issues that we haven't run into, we can always fix them up ourselves. Thanks for the contribution.

**Pull request merged**